### PR TITLE
[DEV-2714] Implement FeatureMaterializeService to prepare features for publishing

### PR DESCRIPTION
**Pull request by YS-L**
## Description

This implements FeatureMaterializeService to prepare features for publishing. It retrieves currently online enabled features that should be materialised together and generates a feature table to be published to offline store. The sql query to generate the feature table is derived from the ingest graphs.

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/service/feature_materialize.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+        if not self.online_store_table_names:
+            return None
+
+        # select distinct serving names across all online store tables
+        distinct_serving_names_from_tables = [
+            select(*[quoted_identifier(serving_name) for serving_name in self.serving_names])
+            .distinct()
+            .from_(table_name)
+            for table_name in self.online_store_table_names
+        ]
```
Comment (gee-senbong) [inline featurebyte/service/feature_materialize.py:105]: Just to double check my understanding, the online store table here refers to offline store table in feast. What we are going to do here is to get the entity universe of all the offline store tables

File: tests/integration/feature_store_integration/test_feature_materialize.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    assert set(materialized_features.names) == set(expected[1:])
+    assert df.shape[0] == 5
+
+    # Check offline store table for combined entity (nothing to materialise here)
+    materialized_features = await service.materialize_features(
+        session=session,
+        primary_entity_ids=[user_entity.id, product_action_entity.id],
+        feature_job_setting=default_feature_job_setting,
+    )
+    assert materialized_features is None
```
Comment (gee-senbong) [inline tests/integration/feature_store_integration/test_feature_materialize.py:143]: If i understood it correctly, if we add a feature that is grouped by `["ÜSER ID", "PRODUCT_ACTION"]`, we should expect some results here, right?

File: featurebyte/service/feature_materialize.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        -------
+        OfflineIngestGraphMetadata
+        """
+        local_query_graph = QueryGraph()
+        output_nodes = []
+        output_column_names = []
+
+        for feature in self.features:
+            offline_ingest_graphs = feature.extract_offline_store_ingest_query_graphs()
+            for offline_ingest_graph in offline_ingest_graphs:
```
Comment (gee-senbong) [inline featurebyte/service/feature_materialize.py:127]: Does it make sense to you if we include `online_store_table_name` in offline ingest query graph? The rationale of doing this is to ease the feast feature view construction.

File: featurebyte/service/feature_materialize.py (hunk 1 above)
Comment (YS-L) [reply featurebyte/service/feature_materialize.py:105]: The online store tables here refer to our current online store tables, not the ones in feast. Currently the materialisation is based on `get_online_features`, which is batch materialisation using precomputed aggregation results.

File: featurebyte/service/feature_materialize.py (hunk 3 above)
Comment (YS-L) [reply featurebyte/service/feature_materialize.py:127]: I was thinking of keeping track of that information in a separate collection such as

```python
class OfflineStoreFeatureTable(FeatureByteBaseDocumentModel):

    name: str
    feature_ids: List[PydanticObjectId]
    primary_entity_ids: List[PydanticObjectId]
    # ... and other attributes
```

which will be updated on each feature's enabling / disabling. But I think including the online store table name in offline ingest query graph also makes sense. Which one do you prefer?

In either case, I think we will still have the above collection to simplify `FeatureMaterializeService` and the orchestration of scheduled tasks for feature publishing.

File: tests/integration/feature_store_integration/test_feature_materialize.py (hunk 2 above)
Comment (YS-L) [reply tests/integration/feature_store_integration/test_feature_materialize.py:143]: Yes, that is right.

**Review by gee-senbong — APPROVED** [review]

**Pull request merged**